File: ios/ImageProcessor/ImageProcessorPlugin.m
//
//  ImageProcessorPlugin.m
//  SeniorDesign
//
//  Created by Nicholas Chu on 3/5/22.
//

// Reference: http://nightlyclosures.com/2018/02/09/writing-a-react-native-ios-module-in-swift
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(ImageProcessorPlugin, NSObject)
RCT_EXTERN_METHOD(makePrediction:(NSString *)path resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

// Reference: https://reactnative.dev/docs/0.62/native-modules-ios#implementing--requiresmainqueuesetup
+ (BOOL)requiresMainQueueSetup
{
  return NO;
}
@end

File: ios/ImageProcessor/ImageProcessorPlugin.swift
//
//  ImageProcessorPlugin.swift
//  SeniorDesign
//
//  Created by Nicholas Chu on 3/5/22.
//

import Foundation

// Reference: http://nightlyclosures.com/2018/02/09/writing-a-react-native-ios-module-in-swift
@objc(ImageProcessorPlugin)
class ImageProcessorPlugin: NSObject {
  
  private var imageModel: ImageModel = ImageModel()
  
  @objc func makePrediction(_ path: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    NSLog("# ImageProcessorPlugin.makePrediction(path=%@)", path)
    
    let formattedPath = String(path.suffix(path.count - 7))
    
    if FileManager.default.fileExists(atPath: formattedPath) {
      guard
        let uiImage = UIImage(contentsOfFile: formattedPath)
      else {
        reject("makePrediction error", "Error loading image from path!", nil)
        return
      }
      
      resolve(imageModel.evaluate(uiImage: uiImage))
    } else {
      reject("makePrediction error", "Image cannot be found!", nil)
    }
  }
}
